linkerd/proxy/http/src/orig_proto.rs

```diff
@@ -76,24 +77,43 @@ where
 
         *req.version_mut() = http::Version::HTTP_2;
 
-        self.inner.call(req).map(|mut res| {
-            debug_assert_eq!(res.version(), http::Version::HTTP_2);
-            let version = if let Some(orig_proto) = res.headers_mut().remove(L5D_ORIG_PROTO) {
-                debug!("downgrading {} response: {:?}", L5D_ORIG_PROTO, orig_proto);
+        UpgradeFuture {
+            version,
+            inner: self.inner.call(req),
+        }
+    }
+}
+
+pub struct UpgradeFuture<F> {
+    version: http::Version,
+    inner: F,
+}
+
+impl<F, B> Future for UpgradeFuture<F>
+where
+    F: Future<Item = http::Response<B>>,
+{
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut res = try_ready!(self.inner.poll());
+        let version = res
+            .headers_mut()
+            .remove(L5D_ORIG_PROTO)
+            .and_then(|orig_proto| {
                 if orig_proto == "HTTP/1.1" {
-                    http::Version::HTTP_11
+                    Some(http::Version::HTTP_11)
                 } else if orig_proto == "HTTP/1.0" {
-                    http::Version::HTTP_10
+                    Some(http::Version::HTTP_10)
                 } else {
-                    warn!("unknown {} header value: {:?}", L5D_ORIG_PROTO, orig_proto);
-                    res.version()
```

Comment:
hmm, should we still warn when the orig_proto header had a weird value? we will record a "Downgrading response to None" event, but since it's at the DEBUG level, it probably won't show up in the logs for most users, and this is probably a sign that something is wrong...